File: notifier.py
"""
Webhook notification module.
Supports: WeChat Work (企业微信), DingTalk (钉钉), Telegram Bot.
Includes retry logic (3 attempts) and notification logging.
"""
import json
import time
import logging
import urllib.request

logger = logging.getLogger(__name__)

MAX_RETRIES = 3
RETRY_DELAY = 2  # seconds between retries


def send_wechat(webhook_url, message):
    """Send message via WeChat Work webhook."""
    payload = json.dumps({
        "msgtype": "text",
        "text": {"content": message}
    }).encode()
    req = urllib.request.Request(webhook_url, data=payload,
                                headers={'Content-Type': 'application/json'})
    urllib.request.urlopen(req, timeout=10)


def send_dingtalk(webhook_url, message):
    """Send message via DingTalk webhook."""
    payload = json.dumps({
        "msgtype": "text",
        "text": {"content": message}
    }).encode()
    req = urllib.request.Request(webhook_url, data=payload,
                                headers={'Content-Type': 'application/json'})
    urllib.request.urlopen(req, timeout=10)


def send_telegram(webhook_url, message):
    """Send message via Telegram Bot API. webhook_url should be:
       https://api.telegram.org/bot<TOKEN>/sendMessage?chat_id=<CHAT_ID>
    """
    payload = json.dumps({
        "text": message,
    }).encode()
    req = urllib.request.Request(webhook_url, data=payload,
                                headers={'Content-Type': 'application/json'})
    urllib.request.urlopen(req, timeout=10)


_SENDERS = {
    'wechat': send_wechat,
    'dingtalk': send_dingtalk,
    'telegram': send_telegram,
}


def _log_notification(instance_name, notify_type, message, success, error_msg='', attempts=1):
    """Write a NotificationLog entry. Must be called inside app context."""
    try:
        from models import db, NotificationLog
        log = NotificationLog(
            instance_name=instance_name,
            notify_type=notify_type,
            message=message[:2000],  # truncate very long messages
            success=success,
            error_message=error_msg[:500] if error_msg else '',
            attempts=attempts,
        )
        db.session.add(log)
        db.session.commit()
    except Exception as e:
        logger.error(f"Failed to log notification: {e}")


def send_alert(notify_type, webhook_url, message, instance_name=''):
    """Dispatch alert to the configured channel with retry (up to 3 attempts).
    Logs every attempt to NotificationLog.
    """
    sender = _SENDERS.get(notify_type)
    if not sender:
        logger.warning(f"Unknown notify type: {notify_type}")
        _log_notification(instance_name, notify_type, message, False, f"Unknown type: {notify_type}")
        return False

    last_error = ''
    for attempt in range(1, MAX_RETRIES + 1):
        try:
            sender(webhook_url, message)
            logger.info(f"Alert sent via {notify_type} (attempt {attempt})")
            _log_notification(instance_name, notify_type, message, True, attempts=attempt)
            return True
        except Exception as e:
            last_error = str(e)
            logger.warning(f"Alert attempt {attempt}/{MAX_RETRIES} failed ({notify_type}): {e}")
            if attempt < MAX_RETRIES:
                time.sleep(RETRY_DELAY)

    logger.error(f"Alert failed after {MAX_RETRIES} attempts ({notify_type}): {last_error}")
    _log_notification(instance_name, notify_type, message, False, last_error, MAX_RETRIES)
    return False

File: internal/service/notifier.go
package service

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/WithZeng/dns-panel/internal/database"
	"github.com/WithZeng/dns-panel/internal/models"
)

const (
	maxRetries = 3
	retryDelay = 2 * time.Second
)

func SendAlert(message, instanceName string) bool {
	var cfg models.AlertConfig
	if err := database.DB.First(&cfg).Error; err != nil || !cfg.Enabled {
		return false
	}

	if cfg.TGBotToken == "" || cfg.TGChatID == "" {
		logNotification(instanceName, "telegram", message, false, "Bot Token 或 Chat ID 未配置", 1)
		return false
	}

	var lastErr string
	for attempt := 1; attempt <= maxRetries; attempt++ {
		if err := sendTelegram(cfg.TGBotToken, cfg.TGChatID, message); err == nil {
			log.Printf("Alert sent via Telegram (attempt %d)", attempt)
			logNotification(instanceName, "telegram", message, true, "", attempt)
			return true
		} else {
			lastErr = err.Error()
			log.Printf("Alert attempt %d/%d failed (telegram): %s", attempt, maxRetries, lastErr)
			if attempt < maxRetries {
				time.Sleep(retryDelay)
			}
		}
	}

	log.Printf("Alert failed after %d attempts (telegram): %s", maxRetries, lastErr)
	logNotification(instanceName, "telegram", message, false, lastErr, maxRetries)
	return false
}

func sendTelegram(botToken, chatID, message string) error {
	url := fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage", botToken)
	payload := map[string]any{
		"chat_id":    chatID,
		"text":       message,
		"parse_mode": "HTML",
	}
	return postJSON(url, payload)
}

func postJSON(url string, payload any) error {
	data, _ := json.Marshal(payload)
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Post(url, "application/json", bytes.NewReader(data))
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return fmt.Errorf("HTTP %d", resp.StatusCode)
	}
	return nil
}

func logNotification(instanceName, notifyType, message string, success bool, errMsg string, attempts int) {
	if len(message) > 2000 {
		message = message[:2000]
	}
	if len(errMsg) > 500 {
		errMsg = errMsg[:500]
	}
	entry := models.NotificationLog{
		InstanceName: instanceName,
		NotifyType:   notifyType,
		Message:      message,
		Success:      success,
		ErrorMessage: errMsg,
		Attempts:     attempts,
	}
	if err := database.DB.Create(&entry).Error; err != nil {
		log.Printf("Failed to log notification: %v", err)
	}
}
